src/main/scala/is/hail/stats/LogisticRegressionModel.scala

```diff
@@ -95,33 +99,84 @@ object LikelihoodRatioStats {
 }
 
 case class LikelihoodRatioStats(b: DenseVector[Double], chi2: Double, p: Double) extends LogisticRegressionStats {
-  def toAnnotation = Annotation(b(0), chi2, p)
+  def toAnnotation = Annotation(b(-1), chi2, p)
 }
 
 
+object FirthTest extends LogisticRegressionTest {
+  def test(X: DenseMatrix[Double], y: DenseVector[Double], nullFit: LogisticRegressionFit): LogisticRegressionTestResultWithFit[FirthStats] = {
+    val m = X.cols
+    val m0 = nullFit.b.length
+    val model = new LogisticRegressionModel(X, y)
+    val nullFitFirth = model.fitFirth(nullFit.b)
+
+    if (nullFitFirth.converged) {
+      val nullFitFirthb = DenseVector.zeros[Double](m)
+      nullFitFirthb(0 until m0) := nullFitFirth.b
+
+      val fitFirth = model.fitFirth(nullFitFirthb)
+
+      val firthStats =
+        if (fitFirth.converged) {
+          val chi2 = 2 * (fitFirth.logLkhd - nullFitFirth.logLkhd)
+          val p = chiSquaredTail(m - m0, chi2)
+
+          Some(new FirthStats(fitFirth.b, chi2, p))
+        } else
+          None
+
+      new LogisticRegressionTestResultWithFit[FirthStats](firthStats, fitFirth)
+    } else
+      new LogisticRegressionTestResultWithFit[FirthStats](None, nullFitFirth)
+  }
+
+  def `type` = TStruct(
+    ("firth", LikelihoodRatioStats.`type`),
+    ("fit", LogisticRegressionFit.`type`))
+}
+
+object FirthStats {
+  def `type`: Type = TStruct(
+    ("beta", TDouble),
+    ("chi2", TDouble),
+    ("pval", TDouble))
+}
+
+case class FirthStats(b: DenseVector[Double], chi2: Double, p: Double) extends LogisticRegressionStats {
+  def toAnnotation = Annotation(b(-1), chi2, p)
+}
+
 
 object ScoreTest extends LogisticRegressionTest {
   def test(X: DenseMatrix[Double], y: DenseVector[Double], nullFit: LogisticRegressionFit): LogisticRegressionTestResult[ScoreStats] = {
+    require(nullFit.score.isDefined && nullFit.fisher.isDefined)
+
     val scoreStats = {
       try {
-        val mu = sigmoid(X * nullFit.b)
+        val m = X.cols
+        val b = DenseVector.zeros[Double](m)
+        val score = DenseVector.zeros[Double](m)
+        val fisher = DenseMatrix.zeros[Double](m, m)
 
-        val r0 = 0 until nullFit.df
-        val r1 = nullFit.df to -1
+        val m0 = nullFit.b.length
+
+        val r0 = 0 until m0
+        val r1 = m0 to -1
```

Comment:
Are you sure this works?  This returns an empty range in Scala (although it does remember the correct start and end.)